Fixes #2158: Let users add products to lists

balaji-ramavathu opened the pull request:
## Description
-> Added menu item "your lists" to mainActivity that displays the product lists.
-> Added a way to display user's product lists and all the related data is stored locally. User can add the 
    product list and the list name is taken through a dialog.
-> User can add the product to the lists from the product page to the existing list or into a new list.
-> User can delete the product from the list by swiping the product right and clicking on the appeared delete icon. Similarly user can delete product lists too.
-> All the data of product lists and their products are stored in the database by using greenDao.
-> If a product from the list is clicked, it goes to the product page displaying the full data of the product 
    (Requires internet to load the product page)
#fixes #2158 
 
 ## Screen-shots, if any
 Adding a product list

![add_list](https://user-images.githubusercontent.com/26673203/52203108-56eba500-2896-11e9-9096-dd60d5f50ccd.gif)

Adding a product to an existing/new list

![add_product](https://user-images.githubusercontent.com/26673203/52203123-65d25780-2896-11e9-936c-95a837f92ec9.gif)

balaji-ramavathu commented:
@teolemon please review :-)

balaji-ramavathu commented:
okay, I'll solve it. I think incrementing greenDao schemaVersion would work.

teolemon commented:
- I would add some way to add the product to the eaten list from the Scan view ("I'm eating this") since you can only add from the browsing view
- I would add a simple CSV export like in History. That way the user can do even more interesting things
- In the "Eaten Products" list, I would add a text like "You'll be able to add product to this list by scanning products, dragging the product card up and tapping on "I've eaten this"  (Scan your first product) button
You'll also be able to add products to this list by searching them and clicking on the + button in the top right of the product page.
- In the other lists: "You'll also be able to add products to this list by searching them and clicking on the + button in the top right of the product page."
- Create new list >> Create a new list
- No / Yes (on create a new list) >> Discard, Save

balaji-ramavathu commented:
Yeah okay, I'll implement them :-)

balaji-ramavathu commented:
Summary of the changes made in the above commit: 
Added a checkBox to the Scan view to take input "I'm eating this" and adding the product to Eaten list accordingly. 
Added CSV export of the products in a specific list.
Added info text and the "Scan First Product" button to product Lists view when the list is empty. Special info text for Eaten list :-).
Mentioned string changes are made.
@teolemon Please review :-)

teolemon commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AddProductOverviewFragment.java:
```diff
@@ -113,6 +114,7 @@
     private static final String PARAM_PURCHASE = "add_purchase_places";
     private static final String PARAM_STORE = "add_stores";
     private static final String PARAM_COUNTRIES = "add_countries";
+    private static final String PARAM_EATING ="eating";
```
```suggestion
    private static final String PARAM_EATING = "eating";
```

teolemon commented on app/src/main/res/values/strings.xml:
```diff
@@ -930,5 +931,21 @@
     <string name="additives">Additives</string>
 
+    <!--used for product_lists implementation-->
+    <string name="add_to_product_lists">Add to Your Lists</string>
```
```suggestion
    <string name="add_to_product_lists">Add to your lists</string>
```

teolemon commented on app/src/main/res/values/strings.xml:
```diff
@@ -930,5 +931,21 @@
+    <string name="add_to_product_lists">Add to Your Lists</string>
+    <string name="txt_create_new_list">Create a new list</string>
+    <string name="txt_add_to_new_list">Add to a new list</string>
+    <string name="txt_eaten_products">Eaten Products</string>
```
```suggestion
    <string name="txt_eaten_products">Eaten products</string>
```

teolemon commented on app/src/main/res/values/strings.xml:
```diff
@@ -930,5 +931,21 @@
+    <string name="txt_products_to_buy">Products to buy</string>
+    <string name="export_csv_your_listed_products">Export CSV</string>
+    <string name="txt_exporting_your_listed_products">Exporting Your Products…</string>
+    <string name="txt_your_listed_products_exported">Your Products exported</string>
```
```suggestion
    <string name="txt_your_listed_products_exported">Your products have been exported</string>
```

teolemon commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/YourListedProducts.java:
```diff
@@ -0,0 +1,329 @@
+        String folder_main = " ";
+        String appname = " ";
+        if ((BuildConfig.FLAVOR.equals("off"))) {
+            folder_main = " Open Food Facts ";
```
Why the extra space before and after ?

balaji-ramavathu replied:
just followed the same pattern as in the exportCSV() of HistoryScanActivity as the functionality is  same :-)

teolemon commented:
@balaji-ramavathu small merge conflict to fix

balaji-ramavathu commented:
Fixed :-)

teolemon commented:
I'm going to merge the branch. Thanks a lot @balaji-ramavathu .
We'll iterate on a few UX points a bit later (the card design in your lists)
